### PR TITLE
21659: Updates typing validation code and logic to support required properties, MAJOR

### PR DESCRIPTION
**Pull request by cademack**
'required' is now a supported attribute of an index within a type definition of an assoc.

Additionally, `values` is no longer an appopriate typing property for (assoc)'s, 'additional_indices' should be used to declare the type of any unspecified indices.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -76,15 +76,16 @@ Return types of methods should be specified by comments on the assoc with the pa
...
 | max_size            | Only applicable when `type` is "list" or "assoc". Value should be an integer.
 | enum                | Only applicable when `type` is "string". A list of possible values for the string
 | min                 | Only applicable when `type` is "number". The minimum value.
 | max                 | Only applicable when `type` is "number". The maximum value.
 | indices             | Only applicable when `type` is "assoc". Value is an assoc of named indices to their expected types.
-| additional_indices  | Only applicable when `type` is "assoc". Value should be  similar to `type` values but is specified for indices that were not named in `indices`. This should only be used if `indices` is used as well.
+| additional_indices  | Only applicable when `type` is "assoc". Value should be  similar to `type` values but is defined for all unspecified indices.
 | optional            | The value of this should be a boolean, and this should only be specified for sub-types. This will be automatically added to the top level of type hints based on the default value of the parameter.
```
Comment (fulpm) [inline CONTRIBUTING.md]: Do we still need `optional` here?

**Review by fulpm — DISMISSED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (cademack) [reply CONTRIBUTING.md]: nope, and I need to add 'required'. Thanks for the comment

**Review by howsoRes — DISMISSED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (cademack) [reply CONTRIBUTING.md]: oh I added required already, just gonna remove optional

**Review by fulpm — APPROVED** [review]

**Review by fulpm — CHANGES_REQUESTED** [review]
Looks like `single_react` still needs an update to `feature_bounds_map` to use `additional_indices`

**Review by fulpm — APPROVED** [review]

**Pull request merged**